Dns search domains

vietj opened the pull request:
Motivation:
The current DNSNameResolver resolves only the hostname and ignore totally the OS search domains. The Java resolver uses the list of search domain provided by the OS and implements hostname resolution using search domains.

Modifications:
- modify the DNSNameResolver algorithm to use search domains when any
- configure the default list of search domains using the sun.net.dns.ResolverConfiguration#searchlist

Result:
The DNSNameResolver resolves properly hostname when search domains are configured.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -84,6 +86,23 @@
     }
 
+    static {
+        String[] searchDomains = new String[0];
```
Move the assignment into the catch block.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -162,6 +187,8 @@ public DnsNameResolver(
         this.optResourceEnabled = optResourceEnabled;
         this.hostsFileEntriesResolver = checkNotNull(hostsFileEntriesResolver, "hostsFileEntriesResolver");
         this.resolveCache = resolveCache;
+        this.searchDomains = checkNotNull(searchDomains, "searchDomains");
```
we will need to call `searchDomains.clone()` as we can not be sure the caller will not modify the array later. Yay for java that not has `const` :(

vietj replied:
given that we know it won't be modified, can't we keep it as is and simply add a comment saying that it should be cloned if it's modified at some point ?
or use an immutable arraylist wrapper ?

normanmaurer replied:
I would prefer to clone it as the user could instance the class directly.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -215,6 +242,14 @@ public long queryTimeoutMillis() {
+
+    int ndots() {
+        return ndots;
+    }
```
make both final

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +408,58 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
+    boolean searchDomains(String hostname) {
+        if (searchDomains.length == 0 || (hostname.length() > 0 && hostname.charAt(hostname.length() - 1) == '.')) {
+            return false;
+        } else {
```
remove else as you return anyway in the if block.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +408,58 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
+    class SingleResolverContext extends DnsNameResolverContext<InetAddress> {
 
-        ctx.resolve();
+        public SingleResolverContext(DnsNameResolver parent, String hostname, DnsCache resolveCache) {
```
remove public

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +408,58 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
+        ctx.resolve(promise);
+    }
+
+    class SingleResolverContext extends DnsNameResolverContext<InetAddress> {
```
final ?

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -472,36 +537,47 @@ private boolean doResolveAllCached(String hostname,
-                                result.add(a);
-                            }
-                        }
+    class ListResolverContext extends DnsNameResolverContext<List<InetAddress>> {
```
final ?

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -472,36 +537,47 @@ private boolean doResolveAllCached(String hostname,
-                            }
-                        }
+    class ListResolverContext extends DnsNameResolverContext<List<InetAddress>> {
+        public ListResolverContext(DnsNameResolver parent, String hostname, DnsCache resolveCache) {
```
remove public

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +98,55 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+        if (trySearchDomain) {
+            final Promise<T> original = promise;
+            promise = new DefaultPromise<T>(parent.executor());
+            FutureListener<T> globalListener = new FutureListener<T>() {
```
no need to assign it to globalListener first. Just directly pass it to `addListener` ?

vietj replied:
yep

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +98,55 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        FutureListener<T> sdListener = new FutureListener<T>() {
```
same as above.

vietj replied:
same as ?

normanmaurer replied:
Just directly pass it to addListener

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -267,17 +311,18 @@ private void onResponseAorAAAA(
     private void onResponseCNAME(
             DnsQuestion question, AddressedEnvelope<DnsResponse, InetSocketAddress> response,
-            Map<String, String> cnames, boolean trace) {
+            Map<String, String> cnames, boolean trace, Promise<T> promise) {
```
from the change it seems like we may miss to notify the promise in some cases.

vietj replied:
can you elaborate ?

normanmaurer replied:
@vietj maybe Im missing something but it seems like we could end up not notify the promise.

vietj replied:
I will check, but (unless I'm wrong) when I did the change of the Promise from field to method parameter, I ended up adding Promise in method and propagating it the call tree. So it means that promise wouldn't have been notified previously too.

normanmaurer replied:
I see... I checked again and it seems like its ok

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -332,12 +377,12 @@ private void onResponseCNAME(
     }
 
-    void tryToFinishResolve() {
+    void tryToFinishResolve(Promise<T> promise) {
```
from the change it seems like we may miss to notify the promise in some cases.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -385,7 +430,7 @@ private boolean gotPreferredAddress() {
     }
 
-    private void finishResolve() {
+    private void finishResolve(Promise<T> promise) {
```
from the change it seems like we may miss to notify the promise in some cases.

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -436,7 +481,10 @@ private void finishResolve() {
-            Class<? extends InetAddress> addressType, List<DnsCacheEntry> resolvedEntries);
+            Class<? extends InetAddress> addressType, List<DnsCacheEntry> resolvedEntries, Promise<T> promise);
+
+    protected abstract DnsNameResolverContext<T> newResolverContext(DnsNameResolver parent, String hostname,
```
can this be package private ?

vietj replied:
this was made protected because the abstract `finishResolve` method was `protected` - I can make it package private, however then it would make sense to do the same for `finishResolve`, WDYT ?

normanmaurer replied:
+1 just make them package private.

normanmaurer commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,209 @@
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+public class SearchDomainTest {
```
@vietj please use 4 spaces and not 2.

vietj replied:
my IDE uses 2 spaces - however it seems smart enough to use 4 when it recognize them in existing classes - not in new classes - will change reformat it of course

normanmaurer commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright 2015 The Netty Project
```
2016

normanmaurer commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,209 @@
+  }
+
+  private TestDnsServer dnsServer;
+  private final EventLoopGroup group = new NioEventLoopGroup(1);
```
should this be done in a `@Before` block ?

vietj replied:
newResolver() is called with different search domains configurations in various tests.

normanmaurer replied:
but you call `shutdownGracefully(...)` in `@After` so wouldn't this mean it is shutdown in between ?

vietj replied:
`shutdownGracefully` is called on the `NioEventLoopGroup` that is used by the `dnsServer`.

But I think I see your point now: if the `dnsServer` is stopped in `@After` then it should be set to null to we are sure to not use a `dnsServer` with a shutdown event loop group.

normanmaurer replied:
+1

normanmaurer commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,209 @@
+    assertEquals(store.getAddress("host2.sub.foo.com"), resolved.getHostAddress());
+  }
+
+  private List<String> asStrings(List<InetAddress> list) {
```
static

normanmaurer commented on resolver-dns/src/test/java/io/netty/resolver/dns/TestDnsServer.java:
```diff
@@ -0,0 +1,297 @@
+/*
+ * Copyright 2015 The Netty Project
```
2016

normanmaurer commented:
@vietj all in all looks great! Just a few comments / questions.

normanmaurer commented:
@vietj also please use 4 spaces everywhere :) Thanks!

vietj commented:
@normanmaurer  feedback addressed

vietj commented:
one test fails but it does not make sense to me when I read the log

vietj commented:
the DNS query should be "host1" and not "host1.foo.com" - since `parent.searchDomains(hostname)` returns false for "host1." string

vietj commented:
can I push "debug" commits to print stuff and see how it goes in the CI ?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -84,6 +86,24 @@
+            searchDomains = list.toArray(new String[list.size()]);
+        } catch (Exception ignore) {
+            // Failed to get the system name search domain list.
+            searchDomains = new String[0];
```
nit: `EmptyArrays.EMPTY_STRINGS`?

vietj replied:
will do thanks, wasn't aware of it.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +410,58 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
-                        return false;
-                    }
-                };
+        System.out.println("doResolveUncached " + hostname);
```
remove

Scottmitch replied:
oops I see you are trying to debug failures ... just a reminder then :)

vietj replied:
yes ... wondering what's going on, will revert the debug commits after

vietj commented:
@normanmaurer found the issue that make the test fail in CI only : DnsNameResolver.hostname delegates to IDN.toASCII that removes the trailing dot and rewrite "host1." as "host1" . http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6894622

any advice on how it should be fixed ?

normanmaurer commented:
@vietj maybe add a workaround linking to the issue ? Like if the last char is an `.`and the output lasts  is not a `.` just append it ?

vietj commented:
I will do that

vietj commented:
it looks like CI is not triggered, any idea why ?

normanmaurer commented:
@vietj it should trigger... it is just busy atm building other PRs

vietj commented:
I think now the PR is fine for me, even though the CI does not seem to check

vietj commented:
ah it seems coming now :-)

netkins commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -68,6 +70,7 @@
     private static final InetAddress LOCALHOST_ADDRESS;
 
     static final InternetProtocolFamily[] DEFAULT_RESOLVE_ADDRESS_TYPES = new InternetProtocolFamily[2];
+    static final String[] DEFAULT_SEACH_DOMAINS;
```
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Make this array "private". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1873)

normanmaurer replied:
+1

normanmaurer replied:
never mind this is a false positive

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -84,6 +87,24 @@
+            List<String> list = (List<String>) nameservers.invoke(instance);
+            searchDomains = list.toArray(new String[list.size()]);
+        } catch (Exception ignore) {
+            // Failed to get the system name search domain list.
```
@vietj do we also want to log some debug message ?

vietj replied:
@normanmaurer it is not done in DnsServerAddresses static block, that's why there is nothing special there too

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +98,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+    private void resolve(Promise<T> promise, boolean trySearchDomain) {
+        if (trySearchDomain) {
+            final Promise<T> original = promise;
+            promise = new DefaultPromise<T>(parent.executor());
```
@vietj use ´parent.executor.newPromise()´. No need to construct `DefaultPromise` explicitly

normanmaurer commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +98,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+                                } else {
+                                    if (count < parent.searchDomains().length) {
+                                        String searchDomain = parent.searchDomains()[count++];
+                                        Promise<T> nextPromise = new DefaultPromise<T>(parent.executor());
```
again use `parent.executor().newPromise()`

normanmaurer commented:
@vietj we are almost there :)

netkins commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -31,6 +31,7 @@
 import io.netty.handler.codec.dns.DnsRecordType;
 import io.netty.handler.codec.dns.DnsResponse;
 import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.DefaultPromise;
```
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Remove this unused import 'io.netty.util.concurrent.DefaultPromise'. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AUselessImportCheck)

vietj replied:
sure

normanmaurer commented:
@slandelle @Scottmitch may want to have a final look ?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +410,57 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
+            if (hostname.charAt(idx) == '.') {
+                dots++;
+            }
+            if (dots >= ndots) {
```
refactor this code so that we only check this condition when `dots` changes

```java
if (hostname.charAt(idx) == '.') {
  if (++dots >= ndots) {
    return false;
  }
}
```

vietj replied:
good idea

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -373,28 +410,57 @@ private static void setSuccess(Promise<InetAddress> promise, InetAddress result)
         }
     }
 
+    boolean searchDomains(String hostname) {
```
can you clarify this method name and/or add javadoc? does this test if `hostName` is a search domain?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
-    protected Promise<T> promise() {
-        return promise;
+    void resolve(Promise<T> promise) {
+        resolve(promise, parent.searchDomains().length > 0 && !hostname.endsWith("."));
```
can we share the first part of `DnsNameResolver.searchDomains` with this method? For example it looks like this logic is doing something very similar to what is in `searchDomains`:

```java
if (searchDomains.length == 0 || (hostname.length() > 0 && hostname.charAt(hostname.length() - 1) == '.')) {
  return false;
}
```

vietj replied:
yep will be addressed in the coming refactor.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolver.java:
```diff
@@ -472,40 +538,56 @@ private boolean doResolveAllCached(String hostname,
-        return IDN.toASCII(inetHost);
+        String hostname = IDN.toASCII(inetHost);
+        // Check for http://bugs.java.com/bugdatabase/view_bug.do?bug_id=6894622
+        if (inetHost.endsWith(".") && !hostname.endsWith(".")) {
```
nit: checking ends with a particular character should be cheaper and is done else where in this PR. Maybe add a method to StringUtil:

```java
public static boolean endsWith(CharSequence s, char c) {
 return s.length() > 0 && s.charAt(s.length() - 1) == c;
}
```

vietj replied:
right

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+                    if (future.isSuccess()) {
+                        original.setSuccess(future.getNow());
+                    } else {
+                        future.addListener(new FutureListener<T>() {
```
`future` will be completed at this point because we are in `operationComplete` ... can we remove this additional layer of listener?

vietj replied:
the listener is reused to listen the various search domains `nextPromise.addListener(this)` until one returns a valid result or all have been tried

Scottmitch replied:
@vietj - can't this be reduced to something like this:

```java
                            final Promise<T> original = promise;
                            promise = parent.executor().newPromise();
                            promise.addListener(new FutureListener<T>() {
                            int count;
                            @Override
                            public void operationComplete(Future<T> future) throws Exception {
                                if (future.isSuccess()) {
                                    original.trySuccess(future.getNow());
                                } else {
                                    if (count < parent.searchDomains().length) {
                                        String searchDomain = parent.searchDomains()[count++];
                                        Promise<T> nextPromise = parent.executor().newPromise();
                                        String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
                                        DnsNameResolverContext<T> nextContext = newResolverContext(parent,
                                            nextHostname, resolveCache);
                                        nextContext.internalResolve(nextPromise);
                                        nextPromise.addListener(this);
                                    } else {
                                        original.tryFailure(future.cause());
                                    }
                                }
                            }
                        });
```

vietj replied:
good catch - yes it can

when you suggested it the first time I was looking at the other promise, not the outer one

> On Jul 7, 2016, at 11:52 PM, Scott Mitchell <notifications@github.com> wrote:
> 
> In resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java <https://github.com/netty/netty/pull/5496#discussion_r69994426>:
> 
> >      }
> >  
> > -    void resolve() {
> > +    private void resolve(Promise<T> promise, boolean trySearchDomain) {
> > +        if (trySearchDomain) {
> > +            final Promise<T> original = promise;
> > +            promise = parent.executor().newPromise();
> > +            promise.addListener(new FutureListener<T>() {
> > +                @Override
> > +                public void operationComplete(Future<T> future) throws Exception {
> > +                    if (future.isSuccess()) {
> > +                        original.setSuccess(future.getNow());
> > +                    } else {
> > +                        future.addListener(new FutureListener<T>() {
> @vietj <https://github.com/vietj> - can't this be reduced to something like this:
> 
>                             final Promise<T> original = promise;
>                             promise = parent.executor().newPromise();
>                             promise.addListener(new FutureListener<T>() {
>                             int count;
>                             @Override
>                             public void operationComplete(Future<T> future) throws Exception {
>                                 if (future.isSuccess()) {
>                                     original.trySuccess(future.getNow());
>                                 } else {
>                                     if (count < parent.searchDomains().length) {
>                                         String searchDomain = parent.searchDomains()[count++];
>                                         Promise<T> nextPromise = parent.executor().newPromise();
>                                         String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
>                                         DnsNameResolverContext<T> nextContext = newResolverContext(parent,
>                                             nextHostname, resolveCache);
>                                         nextContext.internalResolve(nextPromise);
>                                         nextPromise.addListener(this);
>                                     } else {
>                                         original.tryFailure(future.cause());
>                                     }
>                                 }
>                             }
>                         });
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/pull/5496/files/b2995bf1761bcd261a2dee147c5012ef6504100c#r69994426>, or mute the thread <https://github.com/notifications/unsubscribe/AANxilsoozlgSZ_PwlciwI6g0XsWphWbks5qTXUugaJpZM4JEKKu>.
>

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,53 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+                }
+            });
+        }
+        if (parent.searchDomains(hostname)) {
```
I'm not too familiar with the concept so forgive me if this is obvious ... but can you add a comment explaining why if `trySearchDomain` is `false` we still care to check `searchDomains`?

vietj replied:
it's actually a design issue with recursiveness . I will make an update to this that address this and other concerns you raised - please review again after

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -435,8 +477,11 @@ private void finishResolve() {
+            Class<? extends InetAddress> addressType, List<DnsCacheEntry> resolvedEntries, Promise<T> promise);
+
+    abstract DnsNameResolverContext<T> newResolverContext(DnsNameResolver parent, String hostname,
+                                                                    DnsCache resolveCache);
```
nit: fix indentation and line up with `DnsNameResolver parent`?

Scottmitch commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    private TestDnsServer dnsServer;
+    private final EventLoopGroup group = new NioEventLoopGroup(1);
```
initialize this in a `@Before` (or `@BeforeClass`) so its lifetime is well controlled by junit.

Scottmitch commented on resolver-dns/src/test/java/io/netty/resolver/dns/SearchDomainTest.java:
```diff
@@ -0,0 +1,210 @@
+        group.shutdownGracefully();
+    }
+
+    @Test
```
add a timeout to these tests because you have `sync` operations in the test

vietj replied:
I changed the tests to use `await` and not `sync`

Scottmitch commented on resolver-dns/src/test/java/io/netty/resolver/dns/TestDnsServer.java:
```diff
@@ -0,0 +1,297 @@
+
+    private final RecordStore store;
+
+    public TestDnsServer(Set<String> domains) {
```
nti: make constructors package private to match class accessibility.

vietj commented:
@Scottmitch I've addressed concerns (most - not the endsWith yet and not the unit tests) of `DnsNameResolver` and `DnsNameResolverContext`  in a new commit.

I think it should make more sense now in these classes - can you have a look again ?

I will address the other concerns later.

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -135,11 +135,18 @@ public void operationComplete(Future<T> future) throws Exception {
+            int dots = 0;
+            while (idx-- > 0) {
+                if (hostname.charAt(idx) == '.') {
+                    dots++;
```
nit: consider combining this with the previous `if` statement to potentially limit branch instructions

```java
if (hostname.charAt(idx) == '.' && ++dots >= parent.ndoes()) {

}
```

vietj commented:
Pushed the last bits of the feedback from @Scottmitch

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,57 @@ protected DnsNameResolverContext(DnsNameResolver parent,
-        return promise;
+    void resolve(Promise<T> promise) {
+        boolean directSearch = parent.searchDomains().length == 0 ||
+            (hostname.length() > 0 && hostname.charAt(hostname.length() - 1) == '.');
```
plz use the new `StringUtil.endsWith(..)` method

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,56 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+            int dots = 0;
+            while (idx-- > 0) {
+                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
+                    internalResolve(promise);
```
can we move the promise/listener creation and adding the listener into this if block? then below just do `promise.tryFailure(new UnknownHostException(hostname))` (where `promise` is the one in the parameter).

vietj replied:
I don’t follow here - can you clarify please ?

> On Jul 7, 2016, at 11:57 PM, Scott Mitchell <notifications@github.com> wrote:
> 
> In resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java <https://github.com/netty/netty/pull/5496#discussion_r69995158>:
> 
> > +                                        nextContext.internalResolve(nextPromise);
> > +                                        nextPromise.addListener(this);
> > +                                    } else {
> > +                                        original.tryFailure(future.cause());
> > +                                    }
> > +                                }
> > +                            }
> > +                        });
> > +                    }
> > +                }
> > +            });
> > +            int idx = hostname.length();
> > +            int dots = 0;
> > +            while (idx-- > 0) {
> > +                if (hostname.charAt(idx) == '.' && ++dots >= parent.ndots()) {
> > +                    internalResolve(promise);
> can we move the promise/listener creation and adding the listener into this if block? then below just do promise.tryFailure(new UnknownHostException(hostname)) (where promise is the one in the parameter).
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/netty/netty/pull/5496/files/50aca59178618082f484c1190a51a7c41a096c87#r69995158>, or mute the thread <https://github.com/notifications/unsubscribe/AANxirgzvIBm2fS7EWP2FdWTKSuU6ytTks5qTXZXgaJpZM4JEKKu>.
>

Scottmitch replied:
sure. I'll just provide the code bcz it gets the point across :) We don't need to create a new promise, add a listener, etc... if we never find the number of expected dots.

```java
    void resolve(Promise<T> promise) {
        boolean directSearch = parent.searchDomains().length == 0 ||
            (hostname.length() > 0 && hostname.charAt(hostname.length() - 1) == '.');
        if (directSearch) {
            internalResolve(promise);
        } else {
            int dots = 0;
            for (int i = hostname.length() - 1; i >= 0; --i) {
                if (hostname.charAt(i) == '.' && ++dots >= parent.ndots()) {
                    internalResolve(promise);
                    final Promise<T> original = promise;
                    promise = parent.executor().newPromise();
                    promise.addListener(new FutureListener<T>() {
                        @Override
                        public void operationComplete(Future<T> future) throws Exception {
                            if (future.isSuccess()) {
                                original.setSuccess(future.getNow());
                            } else {
                                future.addListener(new FutureListener<T>() {
                                    private int count;
                                    @Override
                                    public void operationComplete(Future<T> future) throws Exception {
                                        if (future.isSuccess()) {
                                            original.trySuccess(future.getNow());
                                        } else {
                                            if (count < parent.searchDomains().length) {
                                                String searchDomain = parent.searchDomains()[count++];
                                                Promise<T> nextPromise = parent.executor().newPromise();
                                                String nextHostname = DnsNameResolverContext.this.hostname + "." + searchDomain;
                                                DnsNameResolverContext<T> nextContext = newResolverContext(parent,
                                                        nextHostname, resolveCache);
                                                nextContext.internalResolve(nextPromise);
                                                nextPromise.addListener(this);
                                            } else {
                                                original.tryFailure(future.cause());
                                            }
                                        }
                                    }
                                });
                            }
                        }
                    });
                    return;
                }
            }

            promise.tryFailure(new UnknownHostException(hostname));
        }
    }
```

Scottmitch replied:
@vietj - i just updated the above code to reflect the current code ... plus also cleaned up some other minor sutff

vietj replied:
@Scottmitch that won't work because the listener code must be triggered whatsoever, i.e the search path must be tried in all case, but we should first try to do an internal resolve if the number of dots of the hostname is >= ndots

two examples of what the actual code does:

1/ _foo.com_ is a search path and we resolve _host1_ then we should try _host1.foo.com_ because _host1_ does not contain any dot

2/ _foo.com_ is a search path and we resolve _host2.sub_ then the following attempts should be made in this order:
- host2.sub
- host2.sub.foo.com

because _host2.sub_ has a dot in it

Scottmitch replied:
sgtm thanks for explaining!

netkins commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -350,13 +386,13 @@ void tryToFinishResolve() {
 
         // We have at least one resolved address or tried CNAME as the last resort..
-        finishResolve();
+        finishResolve(promise);
```
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1066)

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,47 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+                    if (future.isSuccess()) {
+                        original.trySuccess(future.getNow());
+                    } else {
+                        if (count < parent.searchDomains().length) {
```
combine this conditional with the previous to be an `else if`?

Scottmitch commented on resolver-dns/src/main/java/io/netty/resolver/dns/DnsNameResolverContext.java:
```diff
@@ -100,11 +97,47 @@ protected DnsNameResolverContext(DnsNameResolver parent,
+            });
+            int idx = hostname.length();
+            int dots = 0;
+            while (idx-- > 0) {
```
any reason why you don't use a `for` loop here? Indexing seems a bit "unconventional".

```java
for (int i = hostname.length() - 1; i >= 0; --i) {

}
```

Scottmitch commented:
@vietj - Thanks for your patience ... lgtm ... good work! Please squash and we can pull this in.

normanmaurer commented:
+1

vietj commented:
made a new PR https://github.com/netty/netty/pull/5518

vietj commented:
@Scottmitch thanks for your patience too - you squeeze code until the last bit :-)
